grpc-circuitbreaker-utils/src/main/java/org/hypertrace/circuitbreaker/grpcutils/resilience/ResilienceCircuitBreakerInterceptor.java

```diff
@@ -0,0 +1,142 @@
+            .getCircuitBreakerRegistry();
+    this.resilienceCircuitBreakerProvider =
+        new ResilienceCircuitBreakerProvider(
+            resilicenceCircuitBreakerRegistry, resilienceCircuitBreakerConfigMap);
+  }
+
+  public ResilienceCircuitBreakerInterceptor(
+      CircuitBreakerConfiguration<?> circuitBreakerConfiguration,
+      Clock clock,
+      CircuitBreakerRegistry resilicenceCircuitBreakerRegistry,
+      ResilienceCircuitBreakerProvider resilienceCircuitBreakerProvider) {
+    this.circuitBreakerConfiguration = circuitBreakerConfiguration;
+    this.resilienceCircuitBreakerConfigMap =
+        ResilienceCircuitBreakerConfigConverter.getCircuitBreakerConfigs(
+            circuitBreakerConfiguration.getCircuitBreakerThresholdsMap());
+    this.resilicenceCircuitBreakerRegistry = resilicenceCircuitBreakerRegistry;
+    this.resilienceCircuitBreakerProvider = resilienceCircuitBreakerProvider;
+    this.clock = clock;
+  }
+
+  @Override
+  protected boolean isCircuitBreakerEnabled() {
+    return circuitBreakerConfiguration.isEnabled();
+  }
+
+  @Override
+  protected <ReqT, RespT> ClientCall<ReqT, RespT> createInterceptedCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    return new ForwardingClientCall.SimpleForwardingClientCall<>(
+        next.newCall(method, callOptions)) {
+      CircuitBreaker circuitBreaker;
+      String circuitBreakerKey;
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        Instant startTime = clock.instant();
+        // Wrap response listener to track failures
+        Listener<RespT> wrappedListener =
+            wrapListenerWithCircuitBreaker(responseListener, startTime);
+        super.start(wrappedListener, headers);
+      }
+
+      @SuppressWarnings("unchecked")
+      @Override
+      public void sendMessage(ReqT message) {
+        CircuitBreakerConfiguration<ReqT> config =
+            (CircuitBreakerConfiguration<ReqT>) circuitBreakerConfiguration;
+        if (config.getKeyFunction() == null) {
+          log.debug("Circuit breaker will apply to all requests as keyFunction config is not set");
+          circuitBreakerKey = config.getDefaultCircuitBreakerKey();
+        } else {
+          circuitBreakerKey = config.getKeyFunction().apply(RequestContext.CURRENT.get(), message);
+        }
+        circuitBreaker = resilienceCircuitBreakerProvider.getCircuitBreaker(circuitBreakerKey);
+        if (!circuitBreaker.tryAcquirePermission()) {
+          logCircuitBreakerRejection(circuitBreakerKey, circuitBreaker);
+          String rejectionReason =
+              circuitBreaker.getState() == CircuitBreaker.State.HALF_OPEN
+                  ? "Circuit Breaker is HALF-OPEN and rejecting excess requests"
+                  : "Circuit Breaker is OPEN and blocking requests";
+          throw Status.RESOURCE_EXHAUSTED.withDescription(rejectionReason).asRuntimeException();
```

Comment:
So a nice-to-have would be allowing the config to provide an exception builder with this being the default.